lib/Log/Any/Adapter/DERIV.pm

```diff
@@ -140,8 +140,10 @@ sub new {
     # There are other ways of running containers, but for now "in docker? generate JSON"
     # is at least a starting point.
     $self->{in_container} = -r '/.dockerenv';
-    unless($self->{in_container}) {
-        $self->{fh} = path($0 . '.json.log')->opena_utf8 or die 'unable to open log file - ' . $!;
+    my $json_log_file = $self->{json_log_file};
+    $json_log_file = $0 . '.json.log' if(!$json_log_file && !$self->{in_container});
+    if($json_log_file) {
```

Comment:
We only really care to log to a file if its outside a container, so we can simplify as

```suggestion
    unless ($self->{in_container}) {
        my $json_log_file = $self->{json_log_file} ? $self->{json_log_file} : "$0.json.log";
```

That said, relying on `/.dockerenv` to check if we really are in a container isn't good - a better approach is to inspect `/proc/1/cgroup` to check if the process in a container namespace, typically docker, lxc, or kubepods (for k8s.)  See also https://stackoverflow.com/a/23558932/9633558

Reply:
The logic is, if there is a `json_log_file` parameter, then we should create json log file it is either in docker or not in docker.